### PR TITLE
Adding metal_utils for iree_utils

### PR DESCRIPTION
**Pull request by Ranvirsv**
Added metal_utils.py, changes IREE_DEVICE_MAP and IREE_TARGET_MAP to use metal, added a separate case to handel for metal device in compile_uitls.py

Passes 14 test using pytest -k metal, skips 16, xfail 16

### CODE REVIEW

File: shark/iree_utils/metal_utils.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# Copyright 2020 The Nod Team. All rights reserved.
```
Comment (powderluv) [inline shark/iree_utils/metal_utils.py:1]: 2023

File: shark/iree_utils/metal_utils.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    elif any(x in device_name for x in ("AMD", "Radeon")):
+        triple = f"rdna2-unknown-{system_os}"
+    # Intel Targets
+    elif any(x in device_name for x in ("A770", "A750")):
+        triple = f"arc-770-{system_os}"
+
+    # Adreno Targets
+    elif all(x in device_name for x in ("Adreno", "740")):
+        triple = f"adreno-a740-{system_os}"
+
```
Comment (powderluv) [inline shark/iree_utils/metal_utils.py:121]: you can remove everything but the apple targets.

File: shark/iree_utils/metal_utils.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    print(f"Target : {metal_device}")
+    return None
+
+def get_iree_metal_args(device_num=0, extra_args=[]):
+    # res_metal_flag = ["--iree-flow-demote-i64-to-i32"]
+
+    res_metal_flag = []
+    metal_triple_flag = None
+    for arg in extra_args:
+        if "-iree-metal-target-triple=" in arg:
```
Comment (powderluv) [inline shark/iree_utils/metal_utils.py:157]: see https://discord.com/channels/689900678990135345/1097586660495065178/1118622069547077653 

I think you want `--iree-metal-target-platform`

Comment (Ranvirsv) [reply shark/iree_utils/metal_utils.py:157]: Will fix that

File: apps/stable_diffusion/shark_sd.spec [hunk 4]
```diff
@@ -19,6 +19,7 @@ datas += copy_metadata('importlib_metadata')
 datas += copy_metadata('torch-mlir')
 datas += copy_metadata('omegaconf')
 datas += copy_metadata('safetensors')
+datas += copy_metadata('Pillow')
```
Comment (powderluv) [inline apps/stable_diffusion/shark_sd.spec:22]: this is already in tree ?

**Comment by powderluv** [comment]
Please update the branch

File: apps/stable_diffusion/shark_sd.spec (hunk 4 above)
Comment (Ranvirsv) [reply apps/stable_diffusion/shark_sd.spec]: I did a git pull before pushing, realized later I should not have done that. So I got 2 commits from Ean and Abhishek

**Review by powderluv — APPROVED** [review]

**Comment by powderluv** [comment]
Please run lint with black

**Comment by Ranvirsv** [comment]
Wanted to let you know, there are also some changes I was trying for to get stable diffusion running. (i.e. adding checks for metal device and metal flags)

File: shark/iree_utils/metal_utils.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+from os import linesep
+from shark.iree_utils._common import run_cmd
+import iree.runtime as ireert
+from sys import platform
+from shark.iree_utils.vulkan_target_env_utils import get_vulkan_target_env_flag
+
+
+def get_metal_device_name(device_num=0):
+    vulkaninfo_dump, _ = run_cmd("vulkaninfo")
```
Comment (powderluv) [inline shark/iree_utils/metal_utils.py:25]: this wont work. I used 
```
(shark.venv) ➜  web git:(main) ✗ sysctl -a | grep brand             
machdep.cpu.brand_string: Apple M1 Ultra
```

But I think Lei suggested using `--dump_devices`

File: shark/iree_utils/metal_utils.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    if device_name == "" or device_name == [] or device_name is None:
+        metal_device = get_metal_device_name(device_num=device_num)
+    else:
+        metal_device = device_name
+    triple = get_metal_target_triple(metal_device)
+    if triple is not None:
+        print(
+            f"Found metal device {metal_device}. Using metal target triple {triple}"
+        )
+        return f"-iree-metal-target-triple={triple}"
```
Comment (powderluv) [inline shark/iree_utils/metal_utils.py:85]: shouldn't this be `-iree-metal-target-platform=`

**Review by powderluv — CHANGES_REQUESTED** [review]

**Comment by powderluv** [comment]
ok to have the SD fixups too.

File: shark/iree_utils/metal_utils.py (hunk 6 above)
Comment (Ranvirsv) [reply shark/iree_utils/metal_utils.py:85]: sorry, missed that one
Will fix it, and the `vulkaninfo`

**Comment by powderluv** [comment]
Please uninstall the VulkanSDK and vulkaninfo from your device to test :D

**Comment by Ranvirsv** [comment]
I don't know how I can change this, since it returns `--iree-vulkan-target-env`, and I don't think there is a `--iree-metal-target-env`
```
if metal_triple_flag is not None:
        vulkan_target_env = get_vulkan_target_env_flag(meatal_triple_flag)
        res_metal_flag.append(vulkan_target_env)
    return res_metal_flag
```

**Comment by powderluv** [comment]
For compile you still use vulkan

**Review by powderluv — APPROVED** [review]

**Review by powderluv — APPROVED** [review]

**Pull request merged**